Only use EXCEPTION_IDENTITY_PROTECTED in thread_set_exception_ports

bkayes opened the pull request:
#### 234fe91cff8a3dc96d7b2727e617e270dbcec487
<pre>
Only use EXCEPTION_IDENTITY_PROTECTED in thread_set_exception_ports
<a href="https://bugs.webkit.org/show_bug.cgi?id=247635">https://bugs.webkit.org/show_bug.cgi?id=247635</a>
rdar://101307915

Reviewed by Mark Lam.

First party applications will soon be required to use this flavor of
exception, as it is more favorable from a security perspective. This
change adopts that requirement.

* Source/WTF/wtf/threads/Signals.cpp:

Canonical link: <a href="https://commits.webkit.org/260217@main">https://commits.webkit.org/260217@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/78e90e097cee5656a1ff50256241ee9d60220c14

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/107310 "2 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/16366 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/40260 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/116466 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/115910 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/17792 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/7604 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/99465 "Built successfully") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/113074 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/13426 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/96637 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/41043 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/95330 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/28263 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/82812 "Passed tests") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/1/builds/96683 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/9398 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/29616 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/38/builds/96112 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/83/builds/7392 "Built successfully and passed tests") | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/10042 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/84/builds/6482 "Passed tests") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/46/builds/30769 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/15551 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/49197 "Passed tests") | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/37/builds/104947 "Built successfully") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/7063 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/11595 "Built successfully") | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/45/builds/26007 "Passed tests") | 
| | | | | 
<!--EWS-Status-Bubble-End-->

kmiller68 commented on Source/WTF/wtf/threads/Signals.cpp:
```diff
@@ -203,6 +206,157 @@ kern_return_t catch_mach_exception_raise_state_identity(mach_port_t, mach_port_t
+    kern_return_t kr;
+    mach_msg_type_number_t i, thcnt;
+    thread_act_array_t threads;
+    struct thread_identifier_info tiinfo;
```
Ditto, this should probably be `threadIdentifierInfo` or something.

kmiller68 commented on Source/WTF/wtf/threads/Signals.cpp:
```diff
@@ -203,6 +206,157 @@ kern_return_t catch_mach_exception_raise_state_identity(mach_port_t, mach_port_t
+static thread_act_t threadIDGetThreadPort(uint64_t threadID)
+{
+    kern_return_t kr;
+    mach_msg_type_number_t i, thcnt;
```
This isn't standard WebKit style. Typically we'd have declare `i` in the for loop below and we would spell out `threadCount`.

kmiller68 commented on Source/WTF/wtf/threads/Signals.cpp:
```diff
@@ -203,6 +206,157 @@ kern_return_t catch_mach_exception_raise_state_identity(mach_port_t, mach_port_t
+        if (thread != THREAD_NULL) {
+            // clean up all the unnecessary thread ports like a good citizen
+            ASSERT(thread != threads[i]);
+            (void)mach_port_deallocate(mach_task_self(), threads[i]);
```
Why are you casting the result to `(void)`?

bkayes replied:
I was following the convention that I saw in the kernel that does this to suppress "unused result" warnings. I'll remove it

kmiller68 commented on Source/WTF/wtf/threads/Signals.cpp:
```diff
@@ -203,6 +206,157 @@ kern_return_t catch_mach_exception_raise_state_identity(mach_port_t, mach_port_t
+        }
+
+        kr = thread_info(threads[i], THREAD_IDENTIFIER_INFO,
+        (thread_info_t) &tiinfo, &infoCount);
```
This should be `reinterpret_cast<thread_info_t>(& threadIdentifierInfo)` per WebKit style.

kmiller68 commented on Source/WTF/wtf/threads/Signals.cpp:
```diff
@@ -203,6 +206,157 @@ kern_return_t catch_mach_exception_raise_state_identity(mach_port_t, mach_port_t
+    thread_act_t thread = THREAD_NULL;
+
+    task_threads(mach_task_self(), &threads, &thcnt);
+    for (i = 0; i < thcnt; i++) {
```
Why do we have to iterate through every thread in the process? That seems like it shouldn't be necessary...

bkayes replied:
I agree it seems a bit ridiculous, but unfortunately this is the only way that the kernel supports for matching threadID to threadPort. do you think it will be a significant problem? How many threads might webkit have under its control?

kmiller68 replied:
It's hypothetically performance sensitive as this code can be triggered by WebAssembly out of bounds memory access (we reserve a 4GB virtual address space for WebAssembly and only map RW live pages so the CPU handles bounds checking for us). 

It seems like EXCEPTION_STATE_PROTECTED would give the port, which avoids this dance. What's the advantage from a security perspective of EXCEPTION_IDENTITY_PROTECTED?

bkayes replied:
If the performance becomes an issue, we can look into addressing that by adding a new kernel interface or something.

kmiller68 commented on Source/WTF/wtf/threads/Signals.cpp:
```diff
@@ -203,6 +206,157 @@ kern_return_t catch_mach_exception_raise_state_identity(mach_port_t, mach_port_t
+    UNUSED_PARAM(threadID);
+    UNUSED_PARAM(taskIDToken);
+
+    kern_return_t kr;
```
WebKit style doesn't forward declare variables like this. We'd normally declare it at the first assignment unless there's some other compelling reason to have it up front. Also, ditto on naming.

bkayes replied:
On the naming bit, `kr` is the standard convention for this variable, and is already named as such extensively throughout this file. I think it'd be more confusing to use something else in this case.

kmiller68 commented on Source/WTF/wtf/threads/Signals.cpp:
```diff
@@ -203,6 +206,157 @@ kern_return_t catch_mach_exception_raise_state_identity(mach_port_t, mach_port_t
+        return kr;
+    }
+    return KERN_FAILURE;
+}
```
This logic is shared with the implementation of `catch_mach_exception_raise_state` we should split into a helper function.

bkayes replied:
I don't actually think it's a win here, but I'll defer to your advice.

bkayes commented on Source/WTF/wtf/threads/Signals.cpp:
```diff
@@ -203,6 +206,165 @@ kern_return_t catch_mach_exception_raise_state_identity(mach_port_t, mach_port_t
+    return KERN_FAILURE;
+}
+
+static kern_return_t runSignalHandlers(Signal &signal, SigInfo &info, PlatformRegisters& registers, bool &didHandle, mach_msg_type_number_t dataCount, mach_exception_data_t exceptionData)
```
oh i'm dumb, this needs to be above the #ifdef above.

kmiller68 commented on Source/WTF/wtf/threads/Signals.cpp:
```diff
@@ -185,6 +186,8 @@ inline ptrauth_generic_signature_t hashThreadState(const thread_state_t source)
 #endif
 
+
+
```
Nit: remove this whitespace change.

kmiller68 commented on Source/WTF/wtf/threads/Signals.cpp:
```diff
@@ -203,6 +206,165 @@ kern_return_t catch_mach_exception_raise_state_identity(mach_port_t, mach_port_t
+    Signal signal = fromMachException(exceptionType);
+    RELEASE_ASSERT(signal != Signal::Unknown);
+
+    thread = threadIDGetThreadPort(threadID);
```
thread_act_t thread = ...

MenloDorian commented on Source/WTF/wtf/threads/Signals.cpp:
```diff
@@ -203,6 +204,162 @@ kern_return_t catch_mach_exception_raise_state_identity(mach_port_t, mach_port_t
+        }
+
+        kern_return_t kr = thread_info(threads[i], THREAD_IDENTIFIER_INFO,
+        reinterpret_cast<thread_info_t>(&threadIdentifierInfo), &infoCount);
```
Please indent this by 4 spaces.  Alternatively, just keep as part of the previous line, and don't have to break it out.

MenloDorian commented on Source/WTF/wtf/threads/Signals.cpp:
```diff
@@ -203,6 +204,162 @@ kern_return_t catch_mach_exception_raise_state_identity(mach_port_t, mach_port_t
+static thread_act_t threadIDGetThreadPort(uint64_t threadID)
+{
+    mach_msg_type_number_t threadCount;
+    thread_act_array_t threads;
```
I don't see clean up to deallocate this array.  Please fix.

MenloDorian commented on Source/WTF/wtf/threads/Signals.cpp:
```diff
@@ -203,6 +204,162 @@ kern_return_t catch_mach_exception_raise_state_identity(mach_port_t, mach_port_t
+
+#if CPU(X86_64)
+    PlatformRegisters& registers = state.uts.ts64;
+    ASSERT(state.ash.flavor == X86_THREAD_STATE64);
```
Is there a reason you converted this into a Debug only `ASSERT`?  If not, let's make this a `RELEASE_ASSERT` to match prior behavior?

MenloDorian commented on Source/WTF/wtf/threads/Signals.cpp:
```diff
@@ -203,6 +204,162 @@ kern_return_t catch_mach_exception_raise_state_identity(mach_port_t, mach_port_t
+    ASSERT(state.ash.flavor == X86_THREAD_STATE64);
+#elif CPU(X86)
+    PlatformRegisters& registers = state.uts.ts32;
+    ASSERT(state.ash.flavor == X86_THREAD_STATE32);
```
Ditto.

MenloDorian commented on Source/WTF/wtf/threads/Signals.cpp:
```diff
@@ -203,6 +204,162 @@ kern_return_t catch_mach_exception_raise_state_identity(mach_port_t, mach_port_t
+    ASSERT(state.ash.flavor == X86_THREAD_STATE32);
+#elif CPU(ARM64)
+    PlatformRegisters& registers = state.ts_64;
+    ASSERT(state.ash.flavor == ARM_THREAD_STATE64);
```
Ditto.

MenloDorian commented on Source/WTF/wtf/threads/Signals.cpp:
```diff
@@ -203,6 +204,162 @@ kern_return_t catch_mach_exception_raise_state_identity(mach_port_t, mach_port_t
+    ASSERT(state.ash.flavor == ARM_THREAD_STATE64);
+#elif CPU(ARM)
+    PlatformRegisters& registers = state.ts_32;
+    ASSERT(state.ash.flavor == ARM_THREAD_STATE32);
```
Ditto.

MenloDorian commented on Source/WTF/wtf/threads/Signals.cpp:
```diff
@@ -203,6 +204,162 @@ kern_return_t catch_mach_exception_raise_state_identity(mach_port_t, mach_port_t
+#if CPU(ARM64E) && OS(DARWIN)
+        RELEASE_ASSERT(inStateHash == hashThreadState((thread_state_t)&state));
+#endif
+        kr = thread_set_state(thread, flavor, (thread_state_t)&state,
+            stateCount);
```
Don't we also need to `mach_port_deallocate` `thread` here after this last use of it?

MenloDorian reviewed (approved):
r=me with fixes.

Commit:
Only use EXCEPTION_IDENTITY_PROTECTED in thread_set_exception_ports
https://bugs.webkit.org/show_bug.cgi?id=247635
rdar://101307915

Reviewed by Mark Lam.

First party applications will soon be required to use this flavor of
exception, as it is more favorable from a security perspective. This
change adopts that requirement.

* Source/WTF/wtf/threads/Signals.cpp:

Canonical link: https://commits.webkit.org/260217@main
